synapse/lib/types.py

```diff
@@ -80,38 +80,90 @@ def __init__(self, tlib, name, **info):
         self._guid_alias = info.get('alias')
         # TODO figure out what to do about tlib vs core issues
         self._getTufoByProp = getattr(tlib, 'getTufoByProp', None)
+        self._getSubProps = getattr(tlib, 'getSubProps', None)
+        self._getPropNorm = getattr(tlib, 'getPropNorm', None)
 
     def norm(self, valu, oldval=None):
 
+        if isinstance(valu, (list, tuple)):
+            return self._norm_list(valu, oldval)
+
         if not isinstance(valu, str) or len(valu) < 1:
             self._raiseBadValu(valu)
 
+        return self._norm_str(valu, oldval)
+
+    def _norm_str(self, text, oldval=None):
+        text = text.strip()
+        if not text:
+            self._raiseBadValu(text, mesg='No text left after strip().')
+
+        if text[0] == '(':
+            vals, off = s_syntax.parse_list(text)
+            if off != len(text):
+                self._raiseBadValu(text, off=off, vals=vals,
+                                   mesg='List parting for guid type did not consume all of the input text.')
+            return self._norm_list(vals, oldval)
+
         # generate me one.  we dont care.
-        if valu == '*':
+        if text == '*':
             return s_common.guid(), {}
 
-        if valu[0] != '$':
-            retn = valu.lower().replace('-', '')
+        if text[0] != '$':
+            retn = text.lower().replace('-', '')
             if not isguid(retn):
-                self._raiseBadValu(valu)
+                self._raiseBadValu(text)
 
             return retn, {}
 
         if self._guid_alias is None:
-            self._raiseBadValu(valu, mesg='guid resolver syntax used with non-aliased guid')
+            self._raiseBadValu(text, mesg='guid resolver syntax used with non-aliased guid')
 
         if self._getTufoByProp is None:
-            self._raiseBadValu(valu, mesg='guid resolver syntax used with non-cortex tlib')
+            self._raiseBadValu(text, mesg='guid resolver syntax used with non-cortex tlib')
 
         # ( sigh... eventually everything will be a cortex... )
-        node = self._getTufoByProp(self._guid_alias, valu[1:])
+        node = self._getTufoByProp(self._guid_alias, text[1:])
         if node is None:
-            self._raiseBadValu(valu, mesg='no result for guid resolver',
+            self._raiseBadValu(text, mesg='no result for guid resolver',
                                alias=self._guid_alias)
 
         iden = node[1].get(node[1].get('tufo:form'))
         return iden, {}
 
+    def _norm_list(self, valu, oldval=None):
+
+        if not valu:
+            self._raiseBadValu(valu=valu, mesg='No valus present in list to make a guid with')
+
+        subprops = self._getSubProps(self.name)
```

Comment:
this allows for subs to be spit out when the normed value is a form (that has subs).  however, an arbitrary property which is a ptype='guid' will fail here, since the 'guid' type has no subs itself.

Reply:
getPropNorm gets the normalizer from getPropType(pdef.get('ptype')) which would often behave like it only acted on  forms since they are synonymous in many cases.  However, that means that a given DataType instance knows it's *type* name, but not it's *prop* name.  We'll still need a tweak to construct a new DataType instance for each prop, where each DataType instance has explicit knowledge of it's property name (which would naturally include forms).

For example, any form that is ptype=guid would not currently function either.

Reply:
Fixed this